lib/scholar/neighbors/k_nearest_neighbors.ex

```diff
@@ -0,0 +1,401 @@
+defmodule Scholar.Neighbors.KNearestNeighbors do
+  @moduledoc """
+  The K-Nearest Neighbors.
+  """
+  import Nx.Defn
+  require Nx
+
+  @derive {Nx.Container,
+           keep: [:default_num_neighbors, :weights, :num_classes, :p, :task, :metric],
+           containers: [:data, :labels]}
+  defstruct [:data, :labels, :default_num_neighbors, :weights, :num_classes, :p, :task, :metric]
+
+  opts = [
+    num_neighbors: [
+      type: :pos_integer,
+      default: 5,
+      doc: "The number of neighbors to use by default for `k_neighbors` queries"
+    ],
+    num_classes: [
+      type: :pos_integer,
+      required: true,
+      doc: "Number of classes in provided labels"
+    ],
+    weights: [
+      type: {:in, [:uniform, :distance]},
+      default: :uniform,
+      doc: """
+      Weight function used in prediction. Possible values:
+
+        * `:uniform` - uniform weights. All points in each neighborhood are weighted equally.
+
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ],
+    p: [
+      type: {:custom, Scholar.Options, :positive_number, []},
+      default: 2,
+      doc: """
+      Power parameter for the Minkowski metric. When p = 1, this is equivalent
+      to using manhattan_distance ($L_1$), and euclidean_distance ($L_2$) for p = 2.
+      """
+    ],
+    metric: [
+      type: {:in, [:minkowski, :cosine]},
+      default: :minkowski,
+      doc: ~S"""
+      Name of the metric. Possible values:
+      * `:minkowski` - Minkowski metric. By changing value of `:p` parameter we can set Euclidean, Chebyshev, Manhattan or any arbitrary $L_p$ metric.
+
+      * `:cosine` - Cosine metric.
+      """
```

Comment:
What if we merge `metric` and `p`? We could do `metric: :cosine` and `metric: {:minkowski, 1}`?

Reply:
Yes, we could do that. The only thing that convinces me to stay with `:metric` and `:p` separately is that the check with NimbleOptions is simpler for this setup.

Reply:
That's a good point but I think my suggestion has a better user experience and documentation, so please let's go with that!